### ISSUE TITLE
Could not find artifact which is present in .m2/repository 

### ISSUE THREAD

**Issue by kleewho** [technomancy/leiningen#1839]
I have already cached in local .m2 repo library `lib`. That `lib` isn't accessible in any public repository, but only in my company nexus that's mirroring central repo. I don't have access from home to company's nexus, althought on `maven` projects that depends on `lib` I am able to issue `mvn clean install`
without any problems. Unfortunately leiningen can't find this library:

```
lein deps
Could not find artifact com.acme:lib:jar:1.5.2 in central (http://repo1.maven.org/maven2/)
Could not find artifact com.acme:lib:jar:1.5.2 in clojars (https://clojars.org/repo/)
This could be due to a typo in :dependencies or network issues.
If you are behind a proxy, try setting the 'http_proxy' environment variable.
```

This is my project.clj

```clojure
(defproject remote-clontrol "0.1.0-SNAPSHOT"
  :description "microservice for making bookings"
  :min-lein-version "2.0.0"
  :dependencies [[org.clojure/clojure "1.6.0"]
                 [com.acme/lib "1.5.2"]
                 [net.logstash.logback/logstash-logback-encoder "3.6"]]
  :plugins [[lein-ring "0.9.1"]]
  :ring {:handler remote-clontrol.core.handler/app
         :init remote-clontrol.core.handler/init}
  {:dev {:dependencies [[javax.servlet/servlet-api "2.5"]
                        [ring-mock "0.1.5"]]}})
```

and this is the local repo:

```bash
➜  lib  pwd 
/home/user/.m2/repository/com/acme/lib
➜  lib  ls -l 1.5.2 
razem 88
-rw-r--r-- 1 user user 37188 lut 21 10:38 lib-1.5.2.jar
-rw-r--r-- 1 user user    40 gru  1 13:07 lib-1.5.2.jar.sha1
-rw-r--r-- 1 user user  4675 lut 21 10:37 lib-1.5.2.pom
-rw-r--r-- 1 user user    40 gru  1 13:07 lib-1.5.2.pom.sha1
-rw-r--r-- 1 user user 16435 gru  1 13:07 lib-1.5.2-sources.jar
-rw-r--r-- 1 user user    40 gru  1 13:07 lib-1.5.2-sources.jar.sha1
-rw-r--r-- 1 user user   194 gru  1 13:07 _maven.repositories
-rw-r--r-- 1 user user   177 lut 21 10:38 _remote.repositories
```

The difference in times are because I tried to solve this by `mvn install` the `lib`. I tried also `mvn deploy:deploy-file` which I found [here](https://gist.github.com/stuartsierra/3062743). I also tried adding to profile:

```clojure
{:user
 {:local-repo "../../../.m2/repository"
  :checksum :ignore}}
```

What am I doing wrong?

**Comment by kleewho** [comment]
I'm running lein 2.5.1

**Comment by hyPiRion** [comment]
To summarise from IRC: Seems hard to reproduce (I tried 4-5 maven libraries), might be something related to the pom.

**Comment by kleewho** [comment]
I was able to *"solve"*  and pinpoint the problem. If you take a look once again at directory listing then you'll see that there are two files:

```
-rw-r--r-- 1 user user   194 gru  1 13:07 _maven.repositories
-rw-r--r-- 1 user user   177 lut 21 10:38 _remote.repositories
```
When I delete the `_maven.repositories` then lein is able to find dependencies. 
Content of the `_maven.repositories`:

```
#NOTE: This is an internal implementation file, its format can be changed without prior notice.
#Tue Jan 27 15:17:28 CET 2015
lib-1.5.2.jar>nexus-public=
```

Content of the `_remote.repositories`:

```
#NOTE: This is an Aether internal implementation file, its format can be changed without prior notice.
#Tue Jan 27 15:17:26 CET 2015
lib-1.5.2.pom>nexus-public=
lib-1.5.2.jar>nexus-public=
```

I noticed that clojure dependencies doesn't have the file `_remote.repositories` so I assume that this file is created by `mvn` during dependency resolution. That said, it would mean that `lein` creates `_maven.repositories` which would be somewhat misleading.

Any explanation? Any more viable solution? Remembering to delete some stuff in `.m2/repositories` is a little cumbersome.

**Comment by hyPiRion** [comment]
Yeah, this sounds like a caching bug on our side. The invokations of maven I used also generated `_maven.repositories`, but  ended up containing `central`, `clojars` or `` instead of `nexus-public`.

**Comment by kleewho** [comment]
Maybe I should paste it at the beginning. Here's our [settings.xml](https://gist.github.com/kleewho/f68ac1059860e0adf641). The name is from there but I don't know whether it is helpful in any way

**Comment by kleewho** [comment]
If you would guide me a little I would love to get my hands dirty. I wanted to get familiar with clojure and this bug was blocking me for quite some days but if I would get chance to work for a while with someone experienced I think it would be even better for me

**Comment by hyPiRion** [comment]
Yeah – the issue is probably that `_maven.repositories` contains the repository name of where the files could be found, and for some reason it just errors when the name is unknown.

@kleewho, do you mean you want to solve this particular problem? Sure thing, I can probably guide you along on IRC (`#leiningen`) when I'm not at work.

**Comment by kleewho** [comment]
@hyPiRion yes, I meant exactly that. I'll talk to you on IRC

**Comment by kleewho** [comment]
@hyPiRion, to not waste your time I've started to working on it (debugging right now). I will talk to you if when I need additional help.

**Comment by kleewho** [comment]
From aether's `EnhancedLocalRepositoryManager` javadoc:

```java
/**
 * A local repository manager that builds upon the classical Maven 2.0 local repository structure but additionally keeps
 * track of from what repositories a cached artifact was resolved. Resolution of locally cached artifacts will be
 * rejected in case the current resolution request does not match the known source repositories of an artifact, thereby
 * emulating physically separated artifact caches per remote repository.
 * 
 * @author Benjamin Bentmann
 */
```

This behaviour isn't erroneous afterall. It is intended at least for this implementation. Your temporary solution with `profiles.clj` is working. I wonder is there anything else that can be done.

**Comment by kleewho** [comment]
@hyPiRion maybe some kind of switch to use different implementation. Default would be `EnhancedLocalRepositoryManager` and the other choice would be just `LocalRepositoryManager` which should just work in this situation. However I would advise to set it in profile and from there it's not far to just use workaround proposed by you.

**Comment by hyPiRion** [comment]
Hi there,

Sorry for not chiming in yesterday – A couple of personal matters came up.

Thank you for digging into this! I will have a look at the different repository managers and see what's sensible to do or not. I guess `SimpleLocalRepositoryManager` should be sufficient for our use case, so I don't think it's problematic to change it to that unless the performance somehow degrades.

**Comment by kleewho** [comment]
No problem. I was writing with one eye opened so I wouldn't be able to make any conversation :wink: 

I doubt in any performance loss but I can look into this a little if you want. I'd like to get this thing to the finish. 

I meant `SimpleLocalRepositoryManager` but I was to lazy to check actual name so I came with `LocalRepositoryManager`.

**Comment by hyPiRion** [comment]
Yeah, no worries about the naming, I got what you meant :)

Feel free to look into this as deep as you'd like. Happy to take a patch to Lein if you manage to do a fix here (although I guess this is mainly related to pomegranate)

**Comment by kleewho** [comment]
Damn! I want that free leiningen sticker very much!

**Comment by kleewho** [comment]
After few days I'm back on this but I can't reproduce it and I'm kind of mad that it's working

**Comment by kleewho** [comment]
@hyPiRion 
As far as performance goes, there's no worries with using `SimpleLocalRepositoryManager`.

Also I found how it's made in maven:

```java

        if ( request.isUseLegacyLocalRepository() )
        {
            logger.warn( "Disabling enhanced local repository: using legacy is strongly discouraged to ensure"
                + " build reproducibility." );
            try
            {
                session.setLocalRepositoryManager( simpleLocalRepoMgrFactory.newInstance( session, localRepo ) );
            }
            catch ( NoLocalRepositoryManagerException e )
            {

                logger.warn( "Failed to configure legacy local repository: back to default" );
                session.setLocalRepositoryManager( repoSystem.newLocalRepositoryManager( session, localRepo ) );
            }
        }
```

and this log message about *Disabling enhanced local repository...* worries me a little. 

What's more... at the beginning I wrote that it's working with maven for me. In fact the only reason it was working is because in maven I have settings.xml with correct information about repository in our company. That's why even though the repository itself wasn't available to me `EnhancedLocalRepositoryManager` was fine with my local artifacts. This is equivalent of what you proposed with profiles.clj

I could prepare something similar in either pomegranate or even in leiningen but now it's your role to tell me what should I do:
* implement it in pomegranate (if @cemerick will be ok of course)
* implement it somewhere in leiningen (dragging more java objects into some namespace - something that I believe pomegranate should deal with). I'd use fact that pomegranate's user can provide it's own function that produces session `repository-session-fn`.
* fix it by documentation in leiningen explaining why it's working like that and what's the solution

The last option is the best what I could do. I'm not sure that there's plenty of guys that want to use something not available in clojars or central and don't have repository already in somewhere in profiles.

**Comment by hyPiRion** [comment]
@kleewho This is really valuable information, thanks! :)

I think the easiest and sufficient option would be as you suggest: Just document this edge case properly in Leiningen docs. Happy to take a patch for this.

Considering this is the first (?) time this has been reported in lein history, it's presumably not a widespread issue. However, if we were to fix this, I'd think the best would be to fix/ask how to solve/report this directly to Aether/Pomegranate, but it seems nontrivial.

**Comment by cemerick** [comment]
There's a lot here, and my aether memory is rusty; but is the tl;dr that there is some implementation detail in `mvn` that is controlled in `settings.xml` which `lein` can mirror via some `:user` profile setting? If so, then I'd +1 @hyPiRion on simply documenting the optional behaviour.

**Comment by kleewho** [comment]
@hyPiRion About the fix. The only problem I think would be to produce  `SimpleLocalRepositoryManagerFactory` and then call `simpleLocalRepoMgrFactory.newInstance` with the same arguments as they are in [`cemerick.pomegranate.aether/repository-session`](https://github.com/cemerick/pomegranate/blob/master/src/main/clojure/cemerick/pomegranate/aether.clj#L132), but like I wrote just documenting this will be much better.

@cemerick So currently for `mvn` there are two options. First is to set remote repo with correct name which means the same name as when artifacts were downloaded. Second option is to use legacy artifacts resolution (like in version 2 of `mvn`) and then the names of repository are just ignored. AFAIK we can only mirror first option and I think it's enough.

*tl;dr* I'll document this

**Comment by kleewho** [comment]
@hyPiRion I was finally able to find some time to write this. I added this https://github.com/technomancy/leiningen/wiki/Repeatability#corporate-artifact-repository. English is definitely not my native language so it would be good if you could read it and if needed improve it even though I bet it's not native to you either.

**Comment by hyPiRion** [comment]
@kleewho thanks! I fixed some typos, but the content is good. Native English speakers are of course encouraged to look at it as well.

Accepted PR or not, considering your work on tracking down this, you're certainly earned the right for a sticker (Send me an email).

And yeah, the wiki is open without any need for review – I don't think it's possible to not have that option actually. It'd be rather painful if the [Plugin](https://github.com/technomancy/leiningen/wiki/Plugins) page would have to be reviewed every time someone adds a new plugin.

Would you consider this issue closed now, or is there anything else that should be resolved?

**Comment by kleewho** [comment]
Yes, please close the issue. I never been more proud of solving the issue by documenting it

**Comment by hyPiRion** [comment]
Heh, you should be. :) Thanks again for the investigation you did, I really appreciate it.
